File: graph/connected_components/cc.cpp
#include <bits/stdc++.h>
using namespace std;

const int MAX { 200010 };
bitset<MAX> visited;
vector<int> adj[MAX];

void dfs(int u) {
    if(visited[u]) return;
    visited[u] = true;
    for(auto v: adj[u]) dfs(v);
}

int connected_components(int n) {
    int counter = 0;
    visited.reset();
    for(int i = 1; i <= n; i++) {
        if(not visited[i]) {
            counter++;
            dfs(i);
        }
    }
    return counter;
}

File: graph/connected_components/cc.py
adj = {}
visited = set()

def dfs(u):
    if u in visited: return
    visited.add(u)
    for v in adj[u]: dfs(v)
 

def connected_components(n):
    counter = 0
    for i in range(1, n + 1):
        if i not in visited:
            counter += 1
            dfs(i)
    return counter